src/getDataset.js

```diff
@@ -24,67 +24,67 @@ const requestCertainFileType = async (res, req, additional, query) => {
 
 
 /**
- * Fetch dataset in v1 format (meta + tree files).
- * Returns a array of [success {bool}, dataToSend or response code to return]
+ * Fetch dataset in v1 format (meta + tree files), converting to v2 format
+ * before sending as a JSON response to client.
+ * @param {Object} res express response object
+ * @param {string} metaJsonUrl
+ * @param {string} treeJsonUrl
+ * @returns {undefined} if successfully sent dataset
+ * @throws {NotFound} If either of the JSONs were not able to be fetched
+ * @throws {InternalServerError} If the parsing / conversion of the JSONs (into v2 format) failed
  */
-const requestV1Dataset = async (metaJsonUrl, treeJsonUrl) => {
+const sendV1Dataset = async (res, metaJsonUrl, treeJsonUrl) => {
   utils.verbose(`Trying to request v1 datasets at: "${metaJsonUrl}" & "${treeJsonUrl}`);
   let data, datasetJson;
   try {
     data = await Promise.all([utils.fetchJSON(metaJsonUrl), utils.fetchJSON(treeJsonUrl)]);
   } catch (err) {
     utils.warn("Failed to fetch v1 dataset JSONs");
-    return [false, 404]; // Not Found
+    throw new NotFound();
   }
   try {
     datasetJson = auspice.convertFromV1({tree: data[1], meta: data[0]});
   } catch (err) {
     utils.warn("Failed to convert v1 dataset JSONs to v2");
-    return [false, 500]; // Internal Server Error
+    throw new InternalServerError();
   }
   utils.verbose(`Success fetching & converting v1 auspice JSONs. Sending as a single v2 JSON.`);
-  return [true, datasetJson];
+  return res.send(datasetJson);
 };
 
 /**
- * Returns a promise which will fetch the main dataset.
- * The "main" dataset is assumed to be a v2+ (unified) JSON
- * If this request 404s, then we attempt to fetch a v1-version
- * of the dataset (i.e. meta + tree JSONs) and convert it
- * to v2 format for the response.
- * Successful dataset fetches will resolve.
- * If neither the v1 nor the v2 dataset fetch / parse is successful,
- * then the promise will reject.
+ * Attempt to stream the main (v2) auspice dataset. This function (when `await`ed)
+ * will finish streaming before returning.
+ * @param {Object} res express response object
+ * @param {Object} dataset
+ * @returns {undefined} Returns undefined if streaming was successful
+ * @throws {NotFound} Throws if the dataset didn't exist (or the streaming failed)
  */
-const requestMainDataset = async (res, dataset) => {
+const streamMainV2Dataset = async (res, dataset) => {
   const main = dataset.urlFor("main");
-
-  return new Promise((resolve, reject) => {
-    /* try to stream the (v2+) dataset JSON as the response */
-    const req = request
-      .get(main)
-      .on('error', (err) => reject(err))
-      .on("response", async (response) => { // eslint-disable-line consistent-return
-        if (response.statusCode === 200) {
-          utils.verbose(`Successfully streaming ${main}.`);
-          req.pipe(res);
-          return resolve();
```

Comment:
I believe the `resolve()` is getting called before the stream has completed. 
I was able to get `community/jameshadfield/scratch/ncov-demo` to load locally by waiting for the stream to end: 
```javascript
const stream = req.pipe(res);
return stream.on('end', () => resolve());
```

Reply:
Yup. That's my understanding too, and the race condition which resulted would explain the stochastic behaviour I initially saw. My solution in this PR also listens to the `'end'` event before resolving - are there any changes you are recommending for this PR?